### PR TITLE
BAU: Remove unnecessary exceptions from Certificate Chain Validation filter

### PR DESCRIPTION
**Pull request by adityapahuja**
Exceptions are only for error reporting and handling and should not be used for other purposes like control flow.

Authors: @adityapahuja

### CODE REVIEW

**Comment by kornysietsma** [comment]
This seems to be creating some quite large and complex methods - it's good to get rid of exceptions as control flow, but I'm not sure about inlining too much, we still want to keep methods small - see https://ncrcoe.gitbooks.io/java-for-small-teams/content/style/300_keep_methods_small_and_simple.html

**Comment by kornysietsma** [comment]
Sorry, that comment may have been overly negative - I quite like it overall, a definite improvement.

**Comment by vixus0** [comment]
The methods seem fairly self-contained and appropriately named. I think this is just the sort of thing you can't really split up any further without introducing some cognitive overhead. I think it's more important to encapsulate behaviour over minimising LOC per method.

File: saml-metadata-bindings/src/main/java/uk/gov/ida/saml/metadata/CertificateChainValidationFilter.java [hunk 1]
```diff
@@ -64,22 +61,56 @@ public XMLObject filter(@Nullable XMLObject metadata) {
...
+                if (entityDescriptor.getRoleDescriptors().isEmpty()) {
+                    LOG.warn("EntityDescriptor '{}' has empty role descriptor list, metadata will be filtered out", entityDescriptor.getEntityID());
+                    return null;
+                }
             } else if (metadata instanceof EntitiesDescriptor) {
-                processEntityGroup((EntitiesDescriptor) metadata);
+                EntitiesDescriptor entitiesDescriptor = (EntitiesDescriptor) metadata;
+                filterOutUntrustedEntityDescriptors(entitiesDescriptor);
+                if (entitiesDescriptor.getEntityDescriptors().isEmpty()) {
+                    LOG.warn("EntitiesDescriptor '{}' has empty entity descriptor list, metadata will be filtered out");
```
Comment (chrisholmes) [inline saml-metadata-bindings/src/main/java/uk/gov/ida/saml/metadata/CertificateChainValidationFilter.java]: These should probably be errors if we're filtering out the root element

**Pull request merged**